### PR TITLE
refactor: move dispatch functions to ContextsStates class

### PR DESCRIPTION
**Pull request by feloy**
Signed-off-by: Philippe Martin <phmartin@redhat.com>

### What does this PR do?

refactor: move dispatch functions to ContextsStates class

### What issues does this PR fix or reference?

Part of #8775 

### How to test this PR?

- [x] Tests are covering the bug fix or the new feature

### CODE REVIEW

File: packages/main/src/plugin/kubernetes/kubernetes-context-state.ts [hunk 1]
```diff
@@ -52,21 +52,20 @@ import {
 } from '@kubernetes/client-node';
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/kubernetes-context-state.ts:52]: I noticed we have this file named `kubernetes-context-state` and another in the same folder `context-state`, is this expected ?

File: packages/main/src/plugin/kubernetes/contexts-states.ts [hunk 2]
```diff
@@ -135,7 +201,83 @@ export class ContextsStates {
...
+  dispose(): void {
+    this.disposed = true;
+    clearTimeout(this.dispatchContextsGeneralStateTimer);
+    clearTimeout(this.dispatchCurrentContextGeneralStateTimer);
+    for (const timer of this.dispatchCurrentContextResourceTimers.values()) {
+      clearTimeout(timer);
+    }
+  }
+
+  async disposeContext(name: string): Promise<void> {
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/contexts-states.ts]: `disposeContext` is not called when `dispose()` is called ?

File: packages/main/src/plugin/kubernetes/contexts-states.ts [hunk 3]
```diff
@@ -135,7 +201,83 @@ export class ContextsStates {
     update(val);
   }
 
-  async dispose(name: string): Promise<void> {
+  dispatch(options: DispatchOptions): void {
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/contexts-states.ts:225]: Could we have a few line of documentation above, TBH I am not sure to understand what this is doing :'(

File: packages/main/src/plugin/kubernetes/fake-informer.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
+/**********************************************************************
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/fake-informer.ts]: Should this class use `.spec` ? Can we import stuff from one `.spec` file to another `.spec` ?

If we can, we should probably do something like `test-informer.spec.ts`

If we cannot I think we should use `informer-test.ts` rather than `fake`

We use the terminology `Test*` for classes often 

https://github.com/containers/podman-desktop/blob/66734cee5ce7677d84e2565e3da34231a201b96f/packages/renderer/src/lib/pod/KubernetesTerminalService.spec.ts#L28

https://github.com/containers/podman-desktop/blob/72e287a0f3992ebacfdc4893f73a5facd342d8ab/packages/main/src/plugin/color-registry.spec.ts#L36

https://github.com/containers/podman-desktop/blob/fba9eead7078174bf7806e89c5f48c60dd8d0c37/packages/main/src/plugin/directories.spec.ts#L26

File: packages/main/src/plugin/kubernetes/fake-informer.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+interface InformerErrorEvent {
+  delayMs: number;
+  verb: string;
+  error: string;
+}
+
+export const informerStopMock = vi.fn();
+
+export class FakeInformer {
```
Comment (axel7083) [inline packages/main/src/plugin/kubernetes/fake-informer.ts]: Some comment above as this is a dedicated file

File: packages/main/src/plugin/kubernetes/kubernetes-context-state.ts (hunk 1 above)
Comment (feloy) [reply packages/main/src/plugin/kubernetes/kubernetes-context-state.ts:52]: I would like to work on renaming files/classes globally in the latest PR (https://github.com/containers/podman-desktop/issues/8775#issuecomment-2358182299)

File: packages/main/src/plugin/kubernetes/contexts-states.ts (hunk 2 above)
Comment (feloy) [reply packages/main/src/plugin/kubernetes/contexts-states.ts]: `disposeContext` is probably badly named, as it is not disposing system resources, as `dispose` does, but only _deletes_ an entry from the map.

I'm pushing a new commit to rename disposeContext and ContextInformers.dispose with a better name IMO

File: packages/main/src/plugin/kubernetes/fake-informer.ts (hunk 4 above)
Comment (feloy) [reply packages/main/src/plugin/kubernetes/fake-informer.ts]: > Should this class use `.spec` ? Can we import stuff from one `.spec` file to another `.spec` ?
> 

Sonarjs is not happy with: 
```
  0:1  error  Add some tests to this file or delete it  sonarjs/no-empty-test-file
```

**Comment by feloy** [comment]
@axel7083 I have made some changes, and added more documentation

File: packages/main/src/plugin/kubernetes/fake-informer.ts (hunk 4 above)
Comment (axel7083) [reply packages/main/src/plugin/kubernetes/fake-informer.ts]: I would still rather have a basic test, than having this without the `.spec` format :/

File: packages/main/src/plugin/kubernetes/kubernetes-context-state.ts (hunk 1 above)
Comment (axel7083) [reply packages/main/src/plugin/kubernetes/kubernetes-context-state.ts:52]: 👍

**Review by axel7083 — APPROVED** [review]
LGTM

**Review by benoitf — APPROVED** [review]
